File: backend/internal/sanitize/sanitize.go
// Package sanitize provides input sanitization utilities to prevent XSS and injection attacks.
package sanitize

import (
	"html"
	"regexp"
	"strings"
	"unicode"
)

// Common dangerous patterns to remove
var (
	// scriptTagRegex matches script tags and their content
	scriptTagRegex = regexp.MustCompile(`(?i)<script[^>]*>[\s\S]*?</script>`)

	// styleTagRegex matches style tags and their content
	styleTagRegex = regexp.MustCompile(`(?i)<style[^>]*>[\s\S]*?</style>`)

	// eventHandlerRegex matches event handler attributes
	eventHandlerRegex = regexp.MustCompile(`(?i)\s+on\w+\s*=\s*["'][^"']*["']`)

	// htmlTagRegex matches any HTML tag
	htmlTagRegex = regexp.MustCompile(`<[^>]+>`)

	// multipleSpacesRegex matches multiple consecutive spaces
	multipleSpacesRegex = regexp.MustCompile(`\s+`)

	// dangerousProtocolRegex matches dangerous URL protocols
	dangerousProtocolRegex = regexp.MustCompile(`(?i)^\s*(javascript|data|vbscript|file):`)

	// sqlInjectionPatterns contains common SQL injection patterns
	sqlInjectionPatterns = regexp.MustCompile(`(?i)(union\s+select|drop\s+table|insert\s+into|delete\s+from|update\s+.*set|;\s*--)`)

	// emailRegex validates email format
	emailRegex = regexp.MustCompile(`^[^\s@]+@[^\s@]+\.[^\s@]+$`)

	// pathTraversalRegex matches path traversal attempts
	pathTraversalRegex = regexp.MustCompile(`\.\.[\\/]`)

	// nullByteRegex matches null bytes
	nullByteRegex = regexp.MustCompile(`\x00`)
)

// Text sanitizes text input by stripping HTML tags and normalizing whitespace.
// Use for usernames, titles, and single-line text inputs.
func Text(input string) string {
	if input == "" {
		return ""
	}

	// Remove script tags and content
	result := scriptTagRegex.ReplaceAllString(input, "")

	// Remove style tags and content
	result = styleTagRegex.ReplaceAllString(result, "")

	// Remove event handlers
	result = eventHandlerRegex.ReplaceAllString(result, "")

	// Remove all HTML tags
	result = htmlTagRegex.ReplaceAllString(result, "")

	// Decode HTML entities
	result = html.UnescapeString(result)

	// Normalize whitespace
	result = multipleSpacesRegex.ReplaceAllString(result, " ")

	// Trim leading/trailing whitespace
	result = strings.TrimSpace(result)

	return result
}

// HTML escapes HTML special characters for safe display.
// Use when you need to display user input in HTML context without allowing formatting.
func HTML(input string) string {
	if input == "" {
		return ""
	}
	return html.EscapeString(input)
}

// URL sanitizes a URL, blocking dangerous protocols.
// Returns empty string if the URL is potentially dangerous.
func URL(input string) string {
	if input == "" {
		return ""
	}

	trimmed := strings.TrimSpace(input)

	// Check for dangerous protocols
	if dangerousProtocolRegex.MatchString(trimmed) {
		return ""
	}

	// Allow only safe protocols or relative URLs
	lowerURL := strings.ToLower(trimmed)
	safeProtocols := []string{"http://", "https://", "mailto:", "tel:"}

	hasProtocol := false
	for _, proto := range safeProtocols {
		if strings.HasPrefix(lowerURL, proto) {
			hasProtocol = true
			break
		}
	}

	// If URL has a protocol but it's not in our safe list, reject it
	if strings.Contains(lowerURL, "://") && !hasProtocol {
		return ""
	}

	return trimmed
}

// Email sanitizes and validates an email address.
// Returns empty string if the email is invalid.
func Email(input string) string {
	if input == "" {
		return ""
	}

	// Strip HTML and normalize
	sanitized := Text(input)
	sanitized = strings.ToLower(sanitized)

	// Validate email format
	if !emailRegex.MatchString(sanitized) {
		return ""
	}

	return sanitized
}

// Filename sanitizes a filename for safe storage.
// Removes path traversal attempts and dangerous characters.
func Filename(input string) string {
	if input == "" {
		return ""
	}

	result := input

	// Remove path traversal attempts
	result = pathTraversalRegex.ReplaceAllString(result, "")
	result = strings.ReplaceAll(result, "/", "")
	result = strings.ReplaceAll(result, "\\", "")

	// Remove null bytes
	result = nullByteRegex.ReplaceAllString(result, "")

	// Remove control characters
	result = strings.Map(func(r rune) rune {
		if unicode.IsControl(r) {
			return -1 // Remove the character
		}
		return r
	}, result)

	// Keep only safe characters: alphanumeric, dots, hyphens, underscores
	var builder strings.Builder
	for _, r := range result {
		if unicode.IsLetter(r) || unicode.IsDigit(r) || r == '.' || r == '-' || r == '_' {
			builder.WriteRune(r)
		} else {
			builder.WriteRune('_')
		}
	}
	result = builder.String()

	// Prevent empty or dot-only filenames
	if result == "" || result == "." || result == ".." {
		return "unnamed"
	}

	return result
}

// SQLString checks for common SQL injection patterns.
// Returns true if the input appears safe, false if it contains suspicious patterns.
// Note: This is a defense-in-depth measure. Always use parameterized queries!
func SQLString(input string) bool {
	if input == "" {
		return true
	}

	return !sqlInjectionPatterns.MatchString(input)
}

// StripNullBytes removes null bytes from input.
// Null bytes can be used to bypass security filters.
func StripNullBytes(input string) string {
	return nullByteRegex.ReplaceAllString(input, "")
}

// TruncateString truncates a string to the specified length.
// Useful for preventing buffer overflow or storage issues.
func TruncateString(input string, maxLength int) string {
	if len(input) <= maxLength {
		return input
	}
	return input[:maxLength]
}

// NormalizeName normalizes a person's name.
// Capitalizes first letter of each word, removes extra whitespace.
func NormalizeName(input string) string {
	if input == "" {
		return ""
	}

	// First sanitize the input
	sanitized := Text(input)

	// Title case each word
	words := strings.Fields(sanitized)
	for i, word := range words {
		if len(word) > 0 {
			// Capitalize first letter, lowercase rest
			words[i] = strings.ToUpper(string(word[0])) + strings.ToLower(word[1:])
		}
	}

	return strings.Join(words, " ")
}

// Password validates password length and content.
// Returns an error message if invalid, empty string if valid.
// Does NOT sanitize - passwords should be hashed, not filtered.
func Password(input string, minLength, maxLength int) string {
	if len(input) < minLength {
		return "password too short"
	}
	if len(input) > maxLength {
		return "password too long"
	}
	// Check for null bytes which could cause issues
	if strings.Contains(input, "\x00") {
		return "password contains invalid characters"
	}
	return ""
}

File: frontend/app/lib/sanitize.ts
import DOMPurify, { type Config } from "dompurify";

/**
 * Sanitization utilities for XSS protection
 *
 * Uses DOMPurify to sanitize user input and prevent XSS attacks.
 * Apply these functions to user-provided content before rendering or storing.
 */

/**
 * Default DOMPurify configuration for text-only content
 * Strips all HTML tags, keeping only plain text
 */
const TEXT_ONLY_CONFIG: Config = {
  ALLOWED_TAGS: [], // No HTML tags allowed
  ALLOWED_ATTR: [], // No attributes allowed
  KEEP_CONTENT: true, // Keep text content from stripped tags
  RETURN_TRUSTED_TYPE: false, // Return string, not TrustedHTML
};

/**
 * Configuration for basic formatted text (bold, italic, links)
 */
const BASIC_HTML_CONFIG: Config = {
  ALLOWED_TAGS: ["b", "i", "em", "strong", "a", "br", "p", "ul", "ol", "li"],
  ALLOWED_ATTR: ["href", "title", "target", "rel"],
  ALLOW_DATA_ATTR: false,
  ADD_ATTR: ["target"], // Allow target for links
  FORBID_TAGS: ["script", "style", "iframe", "form", "input", "object", "embed"],
  FORBID_ATTR: ["onerror", "onload", "onclick", "onmouseover", "onfocus", "onblur"],
  RETURN_TRUSTED_TYPE: false,
};

/**
 * Configuration for rich text content (more formatting options)
 */
const RICH_TEXT_CONFIG: Config = {
  ALLOWED_TAGS: [
    "b",
    "i",
    "em",
    "strong",
    "a",
    "br",
    "p",
    "ul",
    "ol",
    "li",
    "h1",
    "h2",
    "h3",
    "h4",
    "h5",
    "h6",
    "blockquote",
    "pre",
    "code",
    "span",
    "div",
    "img",
  ],
  ALLOWED_ATTR: ["href", "title", "target", "rel", "src", "alt", "class", "id"],
  ALLOW_DATA_ATTR: false,
  FORBID_TAGS: ["script", "style", "iframe", "form", "input", "object", "embed", "base"],
  FORBID_ATTR: ["onerror", "onload", "onclick", "onmouseover", "onfocus", "onblur", "style"],
  RETURN_TRUSTED_TYPE: false,
};

/**
 * Sanitize text input by stripping all HTML
 * Use for: usernames, titles, single-line inputs
 *
 * @param input - The string to sanitize
 * @returns Sanitized plain text string
 *
 * @example
 * sanitizeText('<script>alert("xss")</script>Hello')
 * // Returns: 'Hello'
 */
export function sanitizeText(input: string | undefined | null): string {
  if (!input) return "";
  const sanitized = DOMPurify.sanitize(input, TEXT_ONLY_CONFIG);
  return sanitized.trim();
}

/**
 * Sanitize HTML with basic formatting allowed
 * Use for: comments, short descriptions
 *
 * @param input - The HTML string to sanitize
 * @returns Sanitized HTML string with basic formatting
 *
 * @example
 * sanitizeBasicHtml('<b>Hello</b><script>alert("xss")</script>')
 * // Returns: '<b>Hello</b>'
 */
export function sanitizeBasicHtml(input: string | undefined | null): string {
  if (!input) return "";
  return DOMPurify.sanitize(input, BASIC_HTML_CONFIG);
}

/**
 * Sanitize rich HTML content
 * Use for: blog posts, long-form content, rich text editor output
 *
 * @param input - The HTML string to sanitize
 * @returns Sanitized HTML string with rich formatting
 *
 * @example
 * sanitizeRichHtml('<h1>Title</h1><script>alert("xss")</script><p>Content</p>')
 * // Returns: '<h1>Title</h1><p>Content</p>'
 */
export function sanitizeRichHtml(input: string | undefined | null): string {
  if (!input) return "";
  return DOMPurify.sanitize(input, RICH_TEXT_CONFIG);
}

/**
 * Sanitize URL to prevent javascript: protocol attacks
 * Use for: href attributes, external links
 *
 * @param url - The URL to sanitize
 * @returns Sanitized URL or empty string if dangerous
 *
 * @example
 * sanitizeUrl('javascript:alert("xss")')
 * // Returns: ''
 *
 * sanitizeUrl('https://example.com')
 * // Returns: 'https://example.com'
 */
export function sanitizeUrl(url: string | undefined | null): string {
  if (!url) return "";

  // Remove leading/trailing whitespace
  const trimmed = url.trim();

  // Check for dangerous protocols
  const lowerUrl = trimmed.toLowerCase();
  const dangerousProtocols = ["javascript:", "data:", "vbscript:", "file:"];

  for (const protocol of dangerousProtocols) {
    if (lowerUrl.startsWith(protocol)) {
      return "";
    }
  }

  // Allow only http, https, mailto, tel protocols
  const allowedProtocols = ["http://", "https://", "mailto:", "tel:"];
  const hasProtocol = allowedProtocols.some((p) => lowerUrl.startsWith(p));

  // If no protocol, assume relative URL (which is safe)
  // If has protocol, verify it's allowed
  if (lowerUrl.includes("://") && !hasProtocol) {
    return "";
  }

  return trimmed;
}

/**
 * Sanitize email address
 * Basic validation and sanitization
 *
 * @param email - The email to sanitize
 * @returns Sanitized email or empty string if invalid
 */
export function sanitizeEmail(email: string | undefined | null): string {
  if (!email) return "";

  const sanitized = sanitizeText(email).toLowerCase();

  // Basic email format check
  const emailRegex = /^[^\s@]+@[^\s@]+\.[^\s@]+$/;
  if (!emailRegex.test(sanitized)) {
    return "";
  }

  return sanitized;
}

/**
 * Escape HTML entities for safe text display
 * Use when you need to display raw text in HTML context
 *
 * @param input - The string to escape
 * @returns HTML-escaped string
 *
 * @example
 * escapeHtml('<script>alert("xss")</script>')
 * // Returns: '&lt;script&gt;alert(&quot;xss&quot;)&lt;/script&gt;'
 */
export function escapeHtml(input: string | undefined | null): string {
  if (!input) return "";

  const escapeMap: Record<string, string> = {
    "&": "&amp;",
    "<": "&lt;",
    ">": "&gt;",
    '"': "&quot;",
    "'": "&#x27;",
    "/": "&#x2F;",
    "`": "&#x60;",
    "=": "&#x3D;",
  };

  return input.replace(/[&<>"'`=/]/g, (char) => escapeMap[char] || char);
}

/**
 * Sanitize object values recursively
 * Useful for sanitizing form data objects before submission
 *
 * @param obj - The object to sanitize
 * @returns Object with all string values sanitized
 */
export function sanitizeObject<T extends Record<string, unknown>>(obj: T): T {
  const result: Record<string, unknown> = {};

  for (const [key, value] of Object.entries(obj)) {
    if (typeof value === "string") {
      result[key] = sanitizeText(value);
    } else if (Array.isArray(value)) {
      result[key] = value.map((item) =>
        typeof item === "string"
          ? sanitizeText(item)
          : typeof item === "object" && item !== null
            ? sanitizeObject(item as Record<string, unknown>)
            : item
      );
    } else if (typeof value === "object" && value !== null) {
      result[key] = sanitizeObject(value as Record<string, unknown>);
    } else {
      result[key] = value;
    }
  }

  return result as T;
}

/**
 * Strip potentially dangerous characters from filenames
 *
 * @param filename - The filename to sanitize
 * @returns Sanitized filename
 */
export function sanitizeFilename(filename: string | undefined | null): string {
  if (!filename) return "";

  // Remove path traversal attempts
  let sanitized = filename.replace(/\.\./g, "").replace(/[/\\]/g, "");

  // Remove null bytes and control characters
  sanitized = sanitized.replace(/[\x00-\x1f\x80-\x9f]/g, "");

  // Keep only alphanumeric, dots, hyphens, underscores
  sanitized = sanitized.replace(/[^a-zA-Z0-9._-]/g, "_");

  // Prevent empty or dot-only filenames
  if (!sanitized || sanitized === "." || sanitized === "..") {
    return "unnamed";
  }

  return sanitized;
}

// Export DOMPurify for advanced use cases
export { DOMPurify };
